File: src/main/kotlin/net/lab0/nebula/reloaded/ui/MandelbrotPanel.java
package net.lab0.nebula.reloaded.ui;

import kotlin.Pair;
import net.lab0.nebula.reloaded.image.ImageCoordinates;
import net.lab0.nebula.reloaded.image.MandelbrotRenderer;
import net.lab0.nebula.reloaded.image.PlanCoordinates;
import net.lab0.nebula.reloaded.image.PlanViewport;
import net.lab0.nebula.reloaded.image.RasterizationContext;
import net.lab0.nebula.reloaded.mandelbrot.Compute;
import net.lab0.nebula.reloaded.mandelbrot.ComputeOptim2;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.swing.JLabel;
import javax.swing.JPanel;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.event.MouseEvent;
import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicReference;

public class MandelbrotPanel extends JPanel {

    private static final Logger log = LoggerFactory.getLogger(MandelbrotPanel.class);
    private static final Object TOKEN = new Object();

    private PlanViewport viewport = createDefaultViewport();

    private Compute compute = new ComputeOptim2();

    private long iterationLimit = 64L;

    private JLabel realValueLabel;
    private JLabel imgValueLabel;
    private JLabel xValueLabel;
    private JLabel yValueLabel;
    private Pair<MouseEvent, MouseEvent> selectionBox = null;


    private final AtomicReference<BufferedImage> lastRenderingRef = new AtomicReference<>();
    /**
     * Event store to tell that an image update event has been received.
     */
    private final BlockingQueue<Object> blockingQueue = new ArrayBlockingQueue<>(1);
    /**
     * Holder for a single thread: the one is charge of checking that an image update request has been received.
     */
    private final Executor imageUpdateWatcher = Executors.newSingleThreadExecutor();

    public MandelbrotPanel() {
        asyncUpdateMandelbrotRendering();
        imageUpdateWatcher.execute(new ImageUpdateWatcher(this, blockingQueue));
        /*
         * We only want to notify.
         * If a value was already present, then notifying again will not change anything.
         */
        blockingQueue.offer(TOKEN); // NOSONAR
    }

    @Override
    protected void paintComponent(Graphics graphics) {
        super.paintComponent(graphics);
        Graphics2D g = (Graphics2D) graphics;
        synchronized (lastRenderingRef) {
            if (lastRenderingRef.get() != null) {
                g.drawRenderedImage(lastRenderingRef.get(), new AffineTransform());
            }
        }
        if (selectionBox != null) {

        }
    }

    /**
     * Adds a flag to tell that the image has to be updated.
     */
    public void asyncUpdateMandelbrotRendering() {
        blockingQueue.offer(TOKEN);
    }

    void updateMandelbrotRendering() {
        if (this.getWidth() == 0 || this.getHeight() == 0) {
            // skip because can't create and image
            return;
        }

        MandelbrotRenderer renderer = new MandelbrotRenderer(viewport);
        BufferedImage image = renderer.render(
            this.getWidth(),
            this.getHeight(),
            iterationLimit,
            compute
        );
        synchronized (lastRenderingRef) {
            lastRenderingRef.set(image);
        }
    }

    void setRealValueLabel(JLabel realValueLabel) {
        this.realValueLabel = realValueLabel;
    }

    void setImgValueLabel(JLabel imgValueLabel) {
        this.imgValueLabel = imgValueLabel;
    }

    void setxValueLabel(JLabel xValueLabel) {
        this.xValueLabel = xValueLabel;
    }

    void setyValueLabel(JLabel yValueLabel) {
        this.yValueLabel = yValueLabel;
    }

    public JLabel getRealValueLabel() {
        return realValueLabel;
    }

    public JLabel getImgValueLabel() {
        return imgValueLabel;
    }

    public JLabel getXValueLabel() {
        return xValueLabel;
    }

    public JLabel getYValueLabel() {
        return yValueLabel;
    }

    public void setSelectionBox(Pair<MouseEvent, MouseEvent> startToEnd) {
        this.selectionBox = startToEnd;
    }

    public void moveImage(@NotNull Pair<MouseEvent, MouseEvent> movement) {
        RasterizationContext context = new RasterizationContext(viewport, getWidth(), getHeight());
        PlanCoordinates oldPosition = context.convert(new ImageCoordinates(movement.getFirst().getX(), movement.getFirst().getY()));
        PlanCoordinates newPosition = context.convert(new ImageCoordinates(movement.getSecond().getX(), movement.getSecond().getY()));
        viewport = viewport.translate(oldPosition.minus(newPosition));
        asyncUpdateMandelbrotRendering();
    }

    public void resetViewport() {
        viewport = createDefaultViewport();
        asyncUpdateMandelbrotRendering();
    }

    private PlanViewport createDefaultViewport() {
        return new PlanViewport(
            new Pair<>(-2, 2), new Pair<>(-2, 2)
        );
    }

    public PlanViewport getViewport() {
        return viewport;
    }
}

File: src/main/kotlin/net/lab0/nebula/reloaded/ui/MandelbrotPanel.kt
package net.lab0.nebula.reloaded.ui

import net.lab0.nebula.reloaded.compute.mandelbrot.ComputeContext
import net.lab0.nebula.reloaded.compute.mandelbrot.MandelbrotComputeEngine
import net.lab0.nebula.reloaded.image.MandelbrotRenderer
import net.lab0.nebula.reloaded.image.RasterizationContext
import net.lab0.nebula.reloaded.image.withAlpha
import net.lab0.nebula.reloaded.tree.PayloadStatus.EDGE
import net.lab0.nebula.reloaded.tree.PayloadStatus.INSIDE
import net.lab0.nebula.reloaded.tree.PayloadStatus.OUTSIDE
import net.lab0.nebula.reloaded.tree.PayloadStatus.UNDEFINED
import net.lab0.nebula.reloaded.tree.TreeNode
import org.slf4j.LoggerFactory
import java.awt.Color
import java.awt.EventQueue
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.event.MouseEvent
import java.awt.geom.AffineTransform
import java.util.concurrent.atomic.AtomicReference
import javax.swing.JLabel

class MandelbrotPanel(computeContextRef: AtomicReference<MandelbrotComputeContext>) :
    FractalPanel(computeContextRef) {

  private var iterationLimit = 512L

  lateinit var realValueLabel: JLabel
  lateinit var imgValueLabel: JLabel
  lateinit var iterationsValueLabel: JLabel
  lateinit var xValueLabel: JLabel
  lateinit var yValueLabel: JLabel

  private var drawFractal = true
  private var drawTree = true

  private var selectionBox: Pair<MouseEvent, MouseEvent>? = null

  private val lastRenderingRef = AtomicReference<RenderingContext>()


  override fun paintComponent(graphics: Graphics) {
    super.paintComponent(graphics)
    val g = graphics as Graphics2D
    val rasterizationContext = getRasterizationContext()

    synchronized(lastRenderingRef) {
      paintLatestFractalRendering(rasterizationContext, g)
    }

    if (drawTree) {
      paintLatestTree(g)
    }
  }

  private fun paintLatestTree(g: Graphics2D) {
    val tooSmallRatio = 32
    val actualViewport = getActualViewport()

    fun TreeNode.isTooSmall() =
        this.position.width < actualViewport.width / tooSmallRatio &&
            this.position.height < actualViewport.height / tooSmallRatio

    val toRender = computeContextRef.get().tree.getNodesBreadthFirst(
        depthFilter = {
          !it.isTooSmall() &&
              it.position.overlaps(actualViewport)
        },
        filter = {
          it.position.overlaps(actualViewport) &&
              (!it.hasChildren() || it.isTooSmall())
        }
    )
    log.debug("Found ${toRender.size} nodes to render")
    renderAreas(g, toRender)
  }

  private fun paintLatestFractalRendering(
      rasterizationContext: RasterizationContext,
      g: Graphics2D
  ) {
    if (drawFractal && lastRenderingRef.get() != null) {
      /**
       * compute the position difference between the last rendering and
       * the current rendering. This is necessary as the repaint may happen
       * between the moment the image moved and the image is recomputed.
       *
       * If we are in such a situation, repaint the part of the image that
       * can be updated while waiting for the next image to be rendered.
       */

      val lastViewport = lastRenderingRef.get().viewport
      val previousCenter = rasterizationContext.convert(lastViewport.center)
      val currentCenter = rasterizationContext.convert(viewport.center)

      val xOffset = previousCenter.x - currentCenter.x
      val yOffset = previousCenter.y - currentCenter.y

      log.debug("Drawing offset: $xOffset;$yOffset")

      val affineTransform = AffineTransform(
          1.0, // no X scaling
          0.0, // no Y shearing
          0.0, // no X shearing
          1.0, // no Y scaling
          xOffset.toDouble(),
          yOffset.toDouble()
      )
      g.drawRenderedImage(lastRenderingRef.get().rendering, affineTransform)
    }
  }

  private fun renderAreas(
      g: Graphics2D,
      nodes: List<TreeNode>
  ) {
    val rasterizationContext = getRasterizationContext()

    nodes.forEach { node ->
      val color = when (node.payload.status) {
        UNDEFINED -> Color(255, 255, 255)
        OUTSIDE -> Color(0, 150, 200)
        INSIDE -> Color(200, 150, 0)
        EDGE -> Color(0, 128, 0)
      }

      val topLeft = rasterizationContext.convert(node.position.topLeft)
      val bottomRight = rasterizationContext.convert(node.position.bottomRight)
      val width = bottomRight.x - topLeft.x
      val height = bottomRight.y - topLeft.y

      paintAsRegular(g, color, topLeft.x, topLeft.y, width, height)
    }
  }

  private fun paintAsRegular(
      g: Graphics2D,
      color: Color,
      x: Int,
      y: Int,
      width: Int,
      height: Int
  ) {
    g.paint = color.withAlpha(0.25)
    g.fillRect(x + 1, y + 1, width - 2, height - 2)
    g.paint = color.withAlpha(0.5)
    g.drawRect(x, y, width - 1, height - 1)
  }

  override fun doRendering() {
    log.debug("Update rendering")
    if (this.width == 0 || this.height == 0) {
      // skip because can't create an image
      return
    }

    val renderingContext = RenderingContext(
        viewport,
        width,
        height,
        iterationLimit,
        computeContextRef.get().computeEngine
    )
    val renderer = MandelbrotRenderer(renderingContext)
    renderer.render()
    synchronized(lastRenderingRef) {
      lastRenderingRef.set(renderingContext)
    }
  }

  fun setSelectionBox(startToEnd: Pair<MouseEvent, MouseEvent>?) {
    this.selectionBox = startToEnd
  }

  fun setComputeEngine(computeEngine: MandelbrotComputeEngine) {
    log.debug("Switching compute engine to $computeEngine")
    object : Thread() {
      override fun run() {
        this@MandelbrotPanel.computeContextRef.set(
            this@MandelbrotPanel.computeContextRef.get()
                .changeComputeEngine(computeEngine)
        )
      }
    }.start()
  }

  fun setShowFractal(enabled: Boolean) {
    this.drawFractal = enabled
    if (this.drawFractal) {
      asyncUpdateRendering()
    }
    else {
      EventQueue.invokeLater {
        this.repaint()
      }
    }
  }

  fun setShowTree(selected: Boolean) {
    this.drawTree = selected
    if (selected) {
      asyncUpdateRendering()
    }
    else {
      EventQueue.invokeLater {
        this.repaint()
      }
    }
  }

  fun setIterationLimit(limit: Long) {
    this.iterationLimit = limit
    log.debug("Set iteration limit to $limit")
    asyncUpdateRendering()
  }

  fun getIterationLimit() = this.iterationLimit

  // TODO extract nodes computing logic somewhere else
  fun computeTreeOnce() {
    Thread {
      computeContextRef.get().tree.getNodesBreadthFirst {
        it.needsCompute()
      }.parallelStream().forEach {
        it.compute()
      }
      EventQueue.invokeLater {
        this@MandelbrotPanel.repaint()
      }
    }.start()
  }

  fun getInEdgeOutSurfaces(): InEdgeOutUndef {
    val surfaces = computeContextRef.get().tree.getNodesBreadthFirst {
      !it.hasChildren()
    }.groupBy {
      it.payload.status
    }.mapValues {
      it.value.map { it.position.surface }.sum()
    }

    return InEdgeOutUndef(
        surfaces[INSIDE] ?: 0.0,
        surfaces[EDGE] ?: 0.0,
        surfaces[OUTSIDE] ?: 0.0,
        surfaces[UNDEFINED] ?: 0.0
    )
  }

  companion object {
    private val log = LoggerFactory.getLogger(MandelbrotPanel::class.java)
  }
}
